ADR036: CloudHSM isolation

samcrang opened the pull request.

chrisfarms commented on docs/architecture/adr/ADR036-hsm-isolation-in-detail.md:
```diff
@@ -0,0 +1,80 @@
+
+We will prevent unauthorised network access to the CloudHSM by:
+
+ - setting each tenant namespace to have a `default-deny` egress
```
> default-deny egress NetworkPolicy

This may not be true for long (discussion: https://github.com/alphagov/gsp/pull/414), due to the problems a blanket default-deny NetworkPolicy have caused. Instead we will switch to having the Istio sidecar "default-deny" all egress traffic to anything not explicitly allowed.

samcrang replied:
yeah I've been keeping an eye on your work on this. I think we'll hold fire on this ADR until we've figured something else. I suspect the general things we'll be doing (restricting network access) will be similar but the concrete implementation will be different.

chrisfarms replied:
ok so after the work in https://github.com/alphagov/gsp/pull/414 the following is the case:

* Istio will prevent routing of egress traffic unless it is explicitly whitelisted as part of the mesh (ie has a `ServiceEntry` or `VirtualService` defined)
* A `GlobalNetworkPolicy` will deny network traffic to HSM from any `Pod` in the cluster not blessed with the `talksToHsm: true` label

samcrang committed:
Add ADR036 for CloudHSM isolation

This ADR documents the already configured and deployed methods we use to
isolate the CloudHSM from neighbouring tenants on the same cluster and
pods that do not require access to an HSM.

samcrang commented:
I think this is ready for review now. I've squashed the commits and taken all the glory for writing it.
